File: src/mvmctl/cli/cp.py
"""CLI — ``mvm cp`` command — copy files between host and microVMs.

Uses a plain Click command (not a Typer app) because ``mvm cp`` has
variable-length positional arguments (1+ sources + 1 destination) that
don't map cleanly to Typer's subcommand model.  Click's ``nargs=-1``
handles this naturally, and its option parser correctly separates flags
like ``--user`` / ``--key`` / ``--force`` from positional paths.
"""

from __future__ import annotations

from typing import TYPE_CHECKING

import click
from rich.progress import (
    BarColumn,
    Progress,
    SpinnerColumn,
    TextColumn,
    TransferSpeedColumn,
)

from mvmctl.utils.cli import handle_errors, mvm_cli

if TYPE_CHECKING:
    from mvmctl.api.cp_operations import CPOperation
    from mvmctl.api.inputs._cp_input import CPInput
else:
    from mvmctl.api import CPInput, CPOperation


@click.command(name="cp")
@click.argument("args", nargs=-1, required=True)
@click.option("--user", "-u", default=None, help="SSH user for VM connections")
@click.option("--key", default=None, help="SSH private key path or name")
@click.option(
    "--force", "-f", is_flag=True, help="Overwrite existing destination files"
)
@handle_errors
def cp_app(
    args: tuple[str, ...],
    user: str | None,
    key: str | None,
    force: bool,
) -> None:
    """Copy files between host and microVMs using tar-over-SSH.

    Uses ``tar`` on both sides — no guest dependencies beyond POSIX-mandated tar.

    Usage:

        # Copy local files to VM
        mvm cp ./myfile.txt my-vm:/root/

        # Copy multiple local files to VM
        mvm cp file1.txt file2.txt file3.txt my-vm:/dst/
        mvm cp *.txt my-vm:/dst/         # Shell glob expands to multiple files

        # Copy file from VM to local
        mvm cp my-vm:/var/log/syslog ./syslog

        # Copy between VMs
        mvm cp vm1:/data/file.txt vm2:/data/

    Path format: use ``vm_name:/remote/path`` for VM paths,
    plain ``/local/path`` for local paths.

    The last positional argument is always the destination. Everything
    before it is a source. Multiple sources only work for host → VM.
    """
    args_list = list(args)

    if len(args_list) < 2:
        mvm_cli.error(
            "At least two arguments required: one or more sources and a destination"
        )
        raise click.Abort()

    *sources, dst = args_list

    inputs = CPInput(sources=sources, dst=dst, user=user, key=key, force=force)

    progress = Progress(
        SpinnerColumn(),
        TextColumn("[progress.description]{task.description}"),
        BarColumn(),
        TransferSpeedColumn(),
        transient=True,
    )

    with progress:
        task = progress.add_task("Copying...", total=None)

        def on_progress(chunk: int) -> None:
            if progress.tasks[0].total is None:
                progress.update(task, total=chunk)
            progress.update(task, advance=chunk)

        result = CPOperation.copy(inputs, on_progress=on_progress)

    if result.is_ok and result.item:
        msg = result.item.get("message", result.message)
        mvm_cli.success(msg)
        return
    elif result.is_ok:
        mvm_cli.success(result.message)
        return
    else:
        mvm_cli.error(result.message or "Copy failed")
        raise SystemExit(1)

File: internal/cli/cp.go
// Package cli — "mvm cp" command — copy files between host and microVMs
package cli

import (
	"fmt"

	"mvmctl/internal/cli/common"
	"mvmctl/pkg/api"
	"mvmctl/pkg/api/inputs"

	"github.com/spf13/cobra"
)

func NewCpCmd(cpAPI api.CPAPI) *cobra.Command {
	var (
		force  bool
		noSync bool
	)

	cmd := &cobra.Command{
		Use:   "cp [sources...] [destination]",
		Short: "Copy files between host and VM",
		Long: `Copy files between host and microVMs using vsock binary frame protocol.

Fast, secure file transfer over Firecracker's built-in vsock device —
no SSH, no guest dependencies, no network setup required.

Usage:

  # Copy local files to VM (directory mode — preserve source name)
  mvm cp ./myfile.txt my-vm:/root/

  # Copy local file to VM with exact destination name (file mode)
  mvm cp ./myfile.txt my-vm:/root/renamed.txt

  # Copy multiple local files to VM (requires directory destination)
  mvm cp file1.txt file2.txt file3.txt my-vm:/dst/
  mvm cp *.txt my-vm:/dst/         # Shell glob expands to multiple files

  # Copy file from VM to local (directory or file mode)
  mvm cp my-vm:/var/log/syslog ./syslog
  mvm cp my-vm:/var/log/syslog ./logs/        # Preserve remote filename

  # Copy between VMs
  mvm cp vm1:/data/file.txt vm2:/data/

Path format: use "vm_name:/remote/path" for VM paths,
plain "/local/path" for local paths.

The last positional argument is always the destination. Everything
before it is a source. For host-to-VM copies, end the destination
with a "/" to preserve the source filename (directory mode), or
omit the trailing "/" to use the exact destination path (file mode).
Multiple sources require a directory destination (trailing "/").`,
		Args:          cobra.ArbitraryArgs,
		SilenceErrors: true,
		RunE: func(cmd *cobra.Command, args []string) error {
			// --- Validation ---
			if len(args) < 2 {
				return fmt.Errorf("at least two arguments required: one or more sources and a destination")
			}

			input := inputs.CPInput{
				Sources: args[:len(args)-1],
				Dest:    args[len(args)-1],
				Force:   force,
				NoSync:  noSync,
			}

			// --- Progress ---
			prog := common.NewProgress()
			prog.Start("Copying...")

			result, cpErr := cpAPI.CPCopy(cmd.Context(), input, func(current, total int64) {
				if total > 0 {
					pct := int(100 * current / total)
					prog.UpdateText(fmt.Sprintf("Copying... %d%%", pct))
				}
			})

			prog.Stop()

			// --- Result handling ---
			if cpErr != nil {
				return cpErr
			}

			// Success path
			msg := result.Message
			if msg == "" {
				msg = "Copy completed"
			}
			common.Cli.Success(msg)
			return nil
		},
	}

	cmd.Flags().BoolVarP(&force, "force", "f", false, "Overwrite existing destination files")
	cmd.Flags().BoolVarP(&noSync, "no-sync", "", false, "Skip final sync() after transfer (faster but risks data loss on VM stop)")

	return cmd
}
